packages/world/src/modules/keysintable/KeysInTableHook.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { IStoreHook } from "@latticexyz/store/src/IStore.sol";
+
+import { KeysInTable } from "./tables/KeysInTable.sol";
+import { UsedKeysIndex } from "./tables/UsedKeysIndex.sol";
+
+/**
+ * Note: if a table with composite keys is used, only the first key is indexed
+ */
+contract KeysInTableHook is IStoreHook {
+  function handleSet(bytes32 tableId, bytes32[] memory key) internal {
+    bytes32 keysHash = keccak256(abi.encode(key));
+
+    // If the key has not yet been set in the table...
+    if (!UsedKeysIndex.getHas(tableId, keysHash)) {
+      uint32 length = KeysInTable.getLength(tableId);
+
+      // Push the key to the list of keys in this table
+      KeysInTable.pushKeys(tableId, key[0]);
+      KeysInTable.setLength(tableId, length + 1);
+
+      // Update the index to avoid duplicating this key in the array
+      UsedKeysIndex.set(tableId, keysHash, true, length);
+    }
+  }
+
+  function onSetRecord(bytes32 table, bytes32[] memory key, bytes memory) public {
+    handleSet(table, key);
+  }
+
+  function onBeforeSetField(bytes32 table, bytes32[] memory key, uint8, bytes memory) public {}
+
+  function onAfterSetField(bytes32 table, bytes32[] memory key, uint8, bytes memory) public {
+    handleSet(table, key);
+  }
+
+  function onDeleteRecord(bytes32 tableId, bytes32[] memory key) public {
+    bytes32 keysHash = keccak256(abi.encode(key));
+    (bool has, uint32 index) = UsedKeysIndex.get(tableId, keysHash);
+
+    // If the key has not been set in the table...
+    if (has) {
+      uint32 length = KeysInTable.getLength(tableId);
+      bytes32 lastKey = KeysInTable.getKeys(tableId)[length - 1];
+
+      // Remove the key from the list of keys in this table
+      KeysInTable.updateKeys(tableId, index, lastKey);
+      KeysInTable.popKeys(tableId);
```

Comment:
let's add a comment here with a link to an issue that tracks that `.pop` should return the popped field so we're reminded to change this once that ships

Reply:
Related to a recent discussion on discord: should we call `deleteRecord` here if `length` was 1 (so is 0 now)? This would slightly reduce garbage data (empty arrays for tables with no keys) on indexers/clients. It's less important here than for `KeysWithValue` though, because there is a much smaller number of tables compared to every possible value in a table.